subsystems/win/basesrv/dosdev.c

```diff
@@ -146,6 +158,317 @@ IsGlobalSymbolicLink(HANDLE LinkHandle,
+{
+    WCHAR Path[8];
+    NTSTATUS Status;
+    BOOLEAN IsGlobal;
+    UNICODE_STRING PathU;
+    HANDLE SymbolicLinkHandle;
+    OBJECT_ATTRIBUTES ObjectAttributes;
+
+    /* Setup our drive path */
+    wcsncpy(Path, L"\\??\\X:", (sizeof(L"\\??\\X:") / sizeof(WCHAR)));
+    Path[4] = DriveLetter + L'A';
+    Path[6] = UNICODE_NULL;
+
+    /* Prepare everything to open the link */
+    RtlInitUnicodeString(&PathU, Path);
+    InitializeObjectAttributes(&ObjectAttributes,
+                               &PathU,
+                               OBJ_CASE_INSENSITIVE,
+                               NULL,
+                               NULL);
+
+    /* Impersonate the caller */
+    if (!CsrImpersonateClient(NULL))
+    {
+        return FALSE;
+    }
+
+    /* Open our drive letter */
+    Status = NtOpenSymbolicLinkObject(&SymbolicLinkHandle,
+                                      SYMBOLIC_LINK_QUERY,
+                                      &ObjectAttributes);
+
+    CsrRevertToSelf();
+
+    if (!NT_SUCCESS(Status))
+    {
+        return FALSE;
+    }
+
+    /* Check whether it's global */
+    Status = IsGlobalSymbolicLink(SymbolicLinkHandle, &IsGlobal);
+    NtClose(SymbolicLinkHandle);
+
+    if (!NT_SUCCESS(Status))
+    {
+        return FALSE;
+    }
+
+    return IsGlobal;
+}
+
+NTSTATUS
+SendWinStationBSM(DWORD Flags,
+                  LPDWORD Recipients,
+                  UINT Message,
+                  WPARAM wParam,
+                  LPARAM lParam)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+NTSTATUS
+BroadcastDriveLetterChange(LONG DriveLetter,
+                           BOOLEAN RemoveDefinition,
+                           PLUID BroadcastLuid)
+{
+    HANDLE hUser32;
+    NTSTATUS Status;
+    UNICODE_STRING User32U;
+    ANSI_STRING ProcedureName;
+    DWORD Recipients, Flags, wParam;
+    LUID SystemLuid = SYSTEM_LUID;
+    BSMINFO Info;
+    DEV_BROADCAST_VOLUME Volume;
+
+    /* We need a broadcast LUID */
+    if (BroadcastLuid == NULL)
+    {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Get the Csr procedure, and keep it forever */
+    if (PBROADCASTSYSTEMMESSAGEEXW == NULL)
+    {
+        hUser32 = 0;
+        RtlInitUnicodeString(&User32U, L"user32");
+        Status = LdrGetDllHandle(NULL, NULL, &User32U, &hUser32);
+        if (hUser32 != 0 && NT_SUCCESS(Status))
+        {
+            RtlInitString(&ProcedureName, "CsrBroadcastSystemMessageExW");
+            Status = LdrGetProcedureAddress(hUser32,
+                                            &ProcedureName,
+                                            0,
+                                            (PVOID *)&PBROADCASTSYSTEMMESSAGEEXW);
+            if (!NT_SUCCESS(Status))
+            {
+                PBROADCASTSYSTEMMESSAGEEXW = NULL;
+            }
+        }
+    }
+
+    /* If we failed to get broadcast procedure, no more actions left */
+    if (PBROADCASTSYSTEMMESSAGEEXW == NULL)
```

Comment:
This is fine here, though I'd personally find it more logical inside the other if.